server/api/api/endpoints/runs.py

```diff
@@ -417,19 +420,79 @@ async def abort_run(
     except ValueError:
         log_and_raise(HTTPStatus.BAD_REQUEST.value, reason="bad JSON body")
 
-    background_task = await run_in_threadpool(
-        server.api.utils.background_tasks.ProjectBackgroundTasksHandler().create_background_task,
-        db_session,
-        project,
-        background_tasks,
-        server.api.crud.Runs().abort_run,
-        mlrun.mlconf.background_tasks.default_timeouts.operations.run_abortion,
-        # args for abort_run
-        db_session,
-        project,
-        uid,
-        iter,
-        run_updates=data,
+    cache_ttl = (
+        float(mlrun.mlconf.background_tasks.default_timeouts.operations.run_abortion)
+        + 10.0
     )
+    async with _abort_run_background_tasks_cache.get_or_create_locked(
+        key=uid,
+        ttl=cache_ttl,
+    ) as cached_background_task:
+        background_task_indicator, created = cached_background_task
+
+        # check if abortion is already in progress
+        if not created:
+            background_task = None
+            try:
+                background_task = await run_in_threadpool(
+                    server.api.utils.background_tasks.ProjectBackgroundTasksHandler().get_background_task,
+                    db_session,
+                    background_task_indicator.background_task_id,
+                    project,
+                )
+            except mlrun.errors.MLRunNotFoundError:
+                pass
+
+            if (
+                background_task
+                and background_task.status.state
+                == mlrun.common.schemas.BackgroundTaskState.running
+            ):
+                logger.warning(
+                    "Abort run background task already in progress",
+                    project=project,
+                    run_uid=uid,
+                    background_task_id=background_task_indicator.background_task_id,
+                )
+                return background_task
+
+            # background task is not running, override it in cache with the same lock acquired,
+            # it will generate a new id for the background task
+            background_task_indicator = await _abort_run_background_tasks_cache.create(
+                uid, cache_ttl, lock=background_task_indicator.lock
+            )
```

Comment:
Not sure I really like this flow - what is the situation in which the background task is not running but is still in the cache? Is it that the cache ttl has not expired yet? My main concern is the reuse of the lock in two objects, which is never a good idea...
In this case, why not add an API to the cache that will `replace()` a task, and will trigger the cleanup of the previous one immediately, while creating a new object with new uid and new lock?

Reply:
good point again, the thought was to keep the run blocked for other requests coming in but now I understand the issue that the object could be expired by then unless locking the whole cache. will give the `replace` a try.